File: proxy-env.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Manage and check proxy settings
# This script manages the proxy server settings using the networksetup utility.

import os
import subprocess
import argparse
import sys

network_ports = [
    {'name': 'wifi',  'interface': 'Wi-Fi'},
    {'name': 'ether', 'interface': 'USB 10/100/1000 LAN'},
    {'name': 'fxz',   'interface': 'FXZ'},
]

network_ports_names = [item["name"] for item in network_ports]

proxy_config = [
    {'name': 'socks',  'url': 'http://localhost:8080/proxy/proxy-socks.pac'},  # nuc.local socks
    {'name': 'squid',  'url': 'http://localhost:8080/proxy/proxy-squid.pac'},  # nuc.local squid
    {'name': 'local',  'url': 'http://localhost:8080/proxy/proxy-local.pac'},  # localhost squid
    {'name': 'office', 'url': 'http://wpad.iiji.jp/proxy.pac'},                # IIJ Office proxy
]

proxy_config_names = [item["name"] for item in proxy_config]

networksetup = '/usr/sbin/networksetup'


def parse_proxy_output(output: str) -> dict:
    """Parse networksetup -getautoproxyurl output into a dict."""
    result = {}
    for line in output.splitlines():
        if ': ' in line:
            key, value = line.split(': ', 1)
            result[key.strip()] = value.strip()
    return result


def print_proxy_info(interface: str, output: str) -> None:
    """Print proxy info for an interface in a consistent format."""
    parsed = parse_proxy_output(output)
    url = parsed.get('URL', '(none)')
    enabled = parsed.get('Enabled', '(unknown)')
    print(f"{interface}:")
    print(f"  URL:     {url}")
    print(f"  Enabled: {enabled}")


def show_port_proxy(port: str) -> None:
    """Show proxy settings for a specific network interface."""
    try:
        result = subprocess.run(
            [networksetup, "-getautoproxyurl", port],
            capture_output=True, check=True
        )
        print_proxy_info(port, result.stdout.decode('utf-8'))
    except subprocess.CalledProcessError as e:
        print(f"error: {e}")
        sys.exit(1)


def show_all_proxy_status() -> None:
    """Show proxy status for all network interfaces and system proxy."""
    for port in network_ports:
        try:
            result = subprocess.run(
                [networksetup, "-getautoproxyurl", port['interface']],
                capture_output=True, check=True
            )
            print_proxy_info(port['interface'], result.stdout.decode('utf-8'))
        except subprocess.CalledProcessError as e:
            print(f"{port['interface']}:")
            print(f"  error: {e}")
        print()

    print("System (scutil):")
    try:
        result = subprocess.run(
            ["scutil", "--proxy"],
            capture_output=True, check=True
        )
        scutil = {}
        for line in result.stdout.decode('utf-8').splitlines():
            if ' : ' in line:
                key, value = line.strip().split(' : ', 1)
                scutil[key] = value
        enabled = scutil.get('ProxyAutoConfigEnable', '0') == '1'
        url = scutil.get('ProxyAutoConfigURLString', '(null)') if enabled else '(null)'
        print(f"  URL:     {url}")
        print(f"  Enabled: {'Yes' if enabled else 'No'}")
    except subprocess.CalledProcessError as e:
        print(f"  error: {e}")


def show_squid_status() -> None:
    """Show squid configuration and process status."""
    squid_conf = "/opt/homebrew/etc/squid.conf"
    print("Squid Conf:")
    try:
        target = os.readlink(squid_conf)
        print(f"  {target}")
    except OSError:
        if os.path.exists(squid_conf):
            print(f"  {squid_conf} (not a symlink)")
        else:
            print("  (not found)")

    print()
    print("Squid process:")
    try:
        result = subprocess.run(["ps", "-ea"], capture_output=True, check=True)
        lines = [
            line for line in result.stdout.decode('utf-8').splitlines()
            if "squid" in line and "grep" not in line and "proxy-env.py" not in line
        ]
        if lines:
            for line in lines:
                print(f"  {line.strip()}")
        else:
            print("  (not running)")
    except subprocess.CalledProcessError as e:
        print(f"  error: {e}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Manage and check proxy settings")
    parser.add_argument("-d", "--debug", action="store_true", help="enable debug output")
    parser.add_argument("-n", "--name", choices=proxy_config_names, help="proxy config name")
    parser.add_argument("action", choices=['on', 'off', 'show', 'squid', 'list'], help="action to perform")
    parser.add_argument("port", nargs='?', help="target network interface (required for on/off)")
    args = parser.parse_args()

    if args.action == 'show':
        if args.port:
            show_port_proxy(args.port)
        else:
            show_all_proxy_status()
        return

    if args.action == 'squid':
        show_squid_status()
        return

    if args.action == 'list':
        width = max(len(p['name']) for p in proxy_config)
        for p in proxy_config:
            print(f"  {p['name']:<{width}}  {p['url']}")
        return

    # on / off require port
    if not args.port:
        print(f"error: port is required for action '{args.action}'")
        sys.exit(1)

    if args.action == 'on':
        if not args.name:
            print("error: --name is required for action 'on'")
            sys.exit(1)

        result = [item for item in proxy_config if item["name"] == args.name]
        if not result:
            print(f"error: proxy '{args.name}' not found")
            sys.exit(1)

        proxy_server = result[0]['url']

        if args.debug:
            print(f"cmd = {networksetup} -setautoproxyurl {args.port} {proxy_server}")

        try:
            subprocess.run(
                [networksetup, "-setautoproxyurl", args.port, proxy_server],
                check=True
            )
        except subprocess.CalledProcessError as e:
            print(f"error: {e}")
            sys.exit(1)

    if args.debug:
        print(f"cmd = {networksetup} -setautoproxystate {args.port} {args.action}")

    try:
        subprocess.run(
            [networksetup, "-setautoproxystate", args.port, args.action],
            check=True
        )
    except subprocess.CalledProcessError as e:
        print(f"error: {e}")
        sys.exit(1)

    # Show result after on/off
    show_port_proxy(args.port)


if __name__ == "__main__":
    main()

File: proxy-env.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# Manage and check proxy settings
# This script manages the proxy server settings using the networksetup utility.
#
# Usage:
#   proxy-env.rb [options] <action> [port]
#
# Actions:
#   on    Enable proxy for specified port (requires --name and port)
#   off   Disable proxy for specified port (requires port)
#   show  Show proxy settings (optionally for a specific port)
#   squid Show squid configuration and process status
#   list  List available proxy configurations
#
# Options:
#   -d, --debug        Enable debug output
#   -n, --name NAME    Proxy config name (required for 'on')
#   -h, --help         Show this help
#

require 'optparse'
require 'open3'

NETWORKSETUP = '/usr/sbin/networksetup'
SQUID_CONF   = '/opt/homebrew/etc/squid.conf'

NETWORK_PORTS = [
  { name: 'wifi',  interface: 'Wi-Fi' },
  { name: 'ether', interface: 'USB 10/100/1000 LAN' },
  { name: 'fxz',   interface: 'FXZ' }
].freeze

PROXY_CONFIG = [
  { name: 'socks',  url: 'http://localhost:8080/proxy/proxy-socks.pac' },
  { name: 'squid',  url: 'http://localhost:8080/proxy/proxy-squid.pac' },
  { name: 'local',  url: 'http://localhost:8080/proxy/proxy-local.pac' },
  { name: 'office', url: 'http://wpad.iiji.jp/proxy.pac' }
].freeze

class ProxyManager
  def initialize(debug: false)
    @debug = debug
  end

  def show(port = nil)
    if port
      show_port_proxy(port)
    else
      show_all_proxy_status
    end
  end

  def enable(port, name)
    config = PROXY_CONFIG.find { |p| p[:name] == name }
    unless config
      warn "error: proxy '#{name}' not found"
      exit 1
    end

    url = config[:url]
    warn "cmd = #{NETWORKSETUP} -setautoproxyurl #{port} #{url}" if @debug

    run_networksetup!('-setautoproxyurl', port, url)
    run_networksetup!('-setautoproxystate', port, 'on')
    show_port_proxy(port)
  end

  def disable(port)
    warn "cmd = #{NETWORKSETUP} -setautoproxystate #{port} off" if @debug
    run_networksetup!('-setautoproxystate', port, 'off')
    show_port_proxy(port)
  end

  def squid_status
    puts 'Squid Conf:'
    begin
      target = File.readlink(SQUID_CONF)
      puts "  #{target}"
    rescue Errno::EINVAL
      # not a symlink
      if File.exist?(SQUID_CONF)
        puts "  #{SQUID_CONF} (not a symlink)"
      else
        puts '  (not found)'
      end
    end

    puts
    puts 'Squid process:'
    out, status = Open3.capture2('ps', '-ea')
    if status.success?
      lines = out.lines.select do |line|
        line.include?('squid') && !line.include?('grep') && !line.include?('proxy-env.rb')
      end
      if lines.empty?
        puts '  (not running)'
      else
        lines.each { |line| puts "  #{line.strip}" }
      end
    else
      warn '  error: failed to run ps'
    end
  end

  def list_configs
    width = PROXY_CONFIG.map { |p| p[:name].length }.max
    PROXY_CONFIG.each do |p|
      puts "  #{p[:name].ljust(width)}  #{p[:url]}"
    end
  end

  private

  def parse_proxy_output(output)
    output.lines.each_with_object({}) do |line, hash|
      key, value = line.split(': ', 2)
      hash[key.strip] = value.strip if key && value
    end
  end

  def print_proxy_info(interface, output)
    parsed = parse_proxy_output(output)
    url     = parsed['URL'] || '(none)'
    enabled = parsed['Enabled'] || '(unknown)'
    puts "#{interface}:"
    puts "  URL:     #{url}"
    puts "  Enabled: #{enabled}"
  end

  def show_port_proxy(port)
    out, status = Open3.capture2(NETWORKSETUP, '-getautoproxyurl', port)
    if status.success?
      print_proxy_info(port, out)
    else
      warn "error: failed to get proxy for #{port}"
      exit 1
    end
  end

  def show_all_proxy_status
    NETWORK_PORTS.each do |port|
      out, status = Open3.capture2(NETWORKSETUP, '-getautoproxyurl', port[:interface])
      if status.success?
        print_proxy_info(port[:interface], out)
      else
        puts "#{port[:interface]}:"
        puts '  error: failed to get proxy info'
      end
      puts
    end

    puts 'System (scutil):'
    out, status = Open3.capture2('scutil', '--proxy')
    if status.success?
      scutil = out.lines.each_with_object({}) do |line, hash|
        key, value = line.strip.split(' : ', 2)
        hash[key] = value if key && value
      end
      enabled = scutil['ProxyAutoConfigEnable'] == '1'
      url     = enabled ? (scutil['ProxyAutoConfigURLString'] || '(null)') : '(null)'
      puts "  URL:     #{url}"
      puts "  Enabled: #{enabled ? 'Yes' : 'No'}"
    else
      warn '  error: failed to run scutil'
    end
  end

  def run_networksetup!(*args)
    _, status = Open3.capture2(NETWORKSETUP, *args)
    unless status.success?
      warn "error: networksetup #{args.join(' ')} failed"
      exit 1
    end
  end
end

ACTIONS = %w[on off show squid list].freeze
PROXY_NAMES = PROXY_CONFIG.map { |p| p[:name] }.freeze

def main
  options = { debug: false, name: nil }

  parser = OptionParser.new do |opts|
    opts.banner = 'Usage: proxy-env.rb [options] <action> [port]'

    opts.on('-d', '--debug', 'Enable debug output') do
      options[:debug] = true
    end

    opts.on('-n', '--name NAME', PROXY_NAMES, "Proxy config name (#{PROXY_NAMES.join(', ')})") do |v|
      options[:name] = v
    end

    opts.on_tail('-h', '--help', 'Show this help') do
      puts opts
      puts
      puts "Actions: #{ACTIONS.join(', ')}"
      exit 0
    end
  end

  begin
    parser.parse!
  rescue OptionParser::InvalidArgument => e
    warn "error: #{e.message}"
    exit 1
  end

  action = ARGV.shift
  unless ACTIONS.include?(action)
    warn action ? "error: unknown action '#{action}'" : 'error: action is required'
    warn "Available actions: #{ACTIONS.join(', ')}"
    exit 1
  end

  port = ARGV.shift
  manager = ProxyManager.new(debug: options[:debug])

  case action
  when 'show'
    manager.show(port)
  when 'squid'
    manager.squid_status
  when 'list'
    manager.list_configs
  when 'on'
    unless port
      warn "error: port is required for action 'on'"
      exit 1
    end
    unless options[:name]
      warn "error: --name is required for action 'on'"
      exit 1
    end
    manager.enable(port, options[:name])
  when 'off'
    unless port
      warn "error: port is required for action 'off'"
      exit 1
    end
    manager.disable(port)
  end
end

main if $PROGRAM_NAME == __FILE__
